worker/src/lib/HeliconeProxyRequest/forwarder.ts

```diff
@@ -130,44 +173,98 @@ export async function proxyForwarder(
-          proxyRequest,
-          cacheSettings.bucketSettings,
-          env.CACHE_KV,
-          cacheSettings.cacheSeed
-        );
-        if (cachedResponse) {
-          ctx.waitUntil(
-            recordCacheHit(
-              cachedResponse.headers,
-              env,
-              new ClickhouseClientWrapper(env),
-              orgData.organizationId
-            )
-          );
-          return cachedResponse;
+    if (latestMessage.role == "user") {
+      const moderationRequest = new Request(
+        "https://api.openai.com/v1/moderations",
+        {
+          method: "POST",
+          headers: proxyRequest.requestWrapper.headers,
+          body: JSON.stringify({
+            input: latestMessage.content,
+          }),
         }
+      );
+
+      const moderationRequestWrapper = await RequestWrapper.create(
+        moderationRequest,
+        env
+      );
+      if (moderationRequestWrapper.error || !moderationRequestWrapper.data) {
+        return responseBuilder.build({
+          body: moderationRequestWrapper.error,
+          status: 500,
+        });
+      }
+
+      const {
+        data: moderationProxyRequest,
+        error: moderationProxyRequestError,
+      } = await new HeliconeProxyRequestMapper(
+        moderationRequestWrapper.data,
+        provider,
+        env
+      ).tryToProxyRequest();
+
+      if (moderationProxyRequestError !== null) {
+        return responseBuilder.build({
+          body: moderationProxyRequestError,
+          status: 500,
+        });
+      }
+
+      const { data: moderationResponse, error: moderationResponseError } =
+        await handleProxyRequest(moderationProxyRequest);
+      if (moderationResponseError != null) {
+        return responseBuilder.build({
+          body: moderationResponseError,
+          status: 500,
+        });
+      }
+
+      const flaggedForModeration = (
+        (await moderationResponse.response.json()) as OpenAIModerationResponse
+      ).results[0].flagged;
+      proxyRequest.flaggedForModeration = flaggedForModeration;
+      ctx.waitUntil(log(moderationResponse.loggable));
+
+      if (flaggedForModeration == true) {
+        moderationResponse.response.headers.forEach((value, key) => {
+          responseBuilder.setHeader(key, value);
+        });
+
+        const responseContent = {
+          body: JSON.stringify({
+            success: false,
+            error: {
+              code: "PROMPT_FLAGGED_FOR_MODERATION",
+              message:
+                "The given prompt was flagged by the OpenAI Moderation endpoint.",
+              details: `See your Helicone request page for more info: https://www.helicone.ai/requests?${moderationProxyRequest.requestId}`,
+            },
+          }),
+          inheritFrom: moderationResponse.response,
+          status: 400,
+        };
+
+        return responseBuilder
+          .setHeader("content-type", "application/json")
+          .build(responseContent);
```

Comment:
This is a lot of code to add within this method. @colegottdank thoughts on moving this to another file? 

Maybe we can create a moderator class and then we can do something like.


```ts
const mod = new Moderator(...my vars)

const moderationResult = mod.moderate();

if (moderationResult.moderated.isModerated === true) {
    return moderationResult.moderationResponse;
}
```